src/plugins/audioplayers/player.c

```diff
@@ -0,0 +1,413 @@
+#include "gst/gst.h"
+#include "gst/gstelementfactory.h"
+#include "gst/gstmessage.h"
+#include "gst/gstsegment.h"
+#include "platformchannel.h"
+#include <flutter-pi.h>
+#include <plugins/audioplayers.h>
+#include <stdio.h>
+
+FILE_DESCR("AudioPlayer::player")
+
+// Private Class functions
+void AudioPlayer_SourceSetup(GstElement* playbin, GstElement* source, GstElement** p_src);
+gboolean AudioPlayer_OnBusMessage(GstBus* bus, GstMessage* message, AudioPlayer* data);
+gboolean AudioPlayer_OnRefresh(AudioPlayer* data);
+
+// Private methods
+void AudioPlayer_SetPlayback(AudioPlayer* self, int64_t seekTo, double rate);
+void AudioPlayer_OnMediaError(AudioPlayer* self, GError* error, gchar* debug);
+void AudioPlayer_OnMediaStateChange(AudioPlayer* self, GstObject* src, GstState* old_state, GstState* new_state);
+void AudioPlayer_OnPositionUpdate(AudioPlayer* self);
+void AudioPlayer_OnDurationUpdate(AudioPlayer* self);
+void AudioPlayer_OnSeekCompleted(AudioPlayer* self);
+void AudioPlayer_OnPlaybackEnded(AudioPlayer* self);
+
+static int on_bus_fd_ready(sd_event_source* s, int fd, uint32_t revents, void* userdata)
+{
+    struct AudioPlayer* player = userdata;
+    GstMessage* msg;
+
+    (void)s;
+    (void)fd;
+    (void)revents;
+
+    /* DEBUG_TRACE_BEGIN(player, "on_bus_fd_ready"); */
+
+    msg = gst_bus_pop(player->bus);
+    if (msg != NULL) {
+        AudioPlayer_OnBusMessage(player->bus, msg, player);
+        gst_message_unref(msg);
+    }
+
+    /* DEBUG_TRACE_END(player, "on_bus_fd_ready"); */
+
+    return 0;
+}
+
+AudioPlayer* AudioPlayer_new(char* playerId, char* channel)
+{
+    AudioPlayer* self = malloc(sizeof(AudioPlayer));
+    if (self == NULL) {
+        return NULL;
+    }
+
+    self->url = NULL;
+    self->_isInitialized = false;
+    self->_isLooping = false;
+    self->_isSeekCompleted = false;
+    self->_playbackRate = 1.0;
+
+    gst_init(NULL, NULL);
+    self->playbin = gst_element_factory_make("playbin", "playbin");
+    if (!self->playbin) {
+        LOG_ERROR("Not all elements could be created");
+    }
+    g_signal_connect(self->playbin, "source-setup", G_CALLBACK(AudioPlayer_SourceSetup), &self->source);
+
+    self->bus = gst_element_get_bus(self->playbin);
+
+    // Watch bus messages for one time events
+    // gst_bus_add_watch(self->bus, (GstBusFunc)AudioPlayer_OnBusMessage, self);
+
+    GPollFD fd;
+    sd_event_source* busfd_event_source;
```

Comment:
If feasible, declarations should be at the start of the block. That helps with readability IMO